PSEUDO PR - DO NOT MERGE

ewels opened the pull request:
Continuation of https://github.com/nf-core/hicar/pull/1 which was merged.

---

# PSEUDO PR - DO NOT MERGE

This is the pseudo-PR that we use for the community review leading up to the first release of this nf-core pipeline.

***Please do not merge the PR***, as that closes it and all associated discussion. However, you can use the GitHub reviewing interface to add reviews and inline comments to the code.

---

# To do before first release

Bringing over my comments from my initial review in #1, here are the topics that we need to resolve before initial release:

1. [ ] At least one more review from someone else, preferably two
2. [x] Resolve questions about software packaging for Juicer
3. [x] Resolve questions about installation of R packages
4. [ ] Test running the pipeline offline

## Juicer packaging

The pipeline currently contains a [35MB `.jar` file](https://github.com/nf-core/hicar/blob/master/bin/juicer_tools_1.22.01.jar) with Juicer tools. I'm guessing that this is a tricky tool to package or something, I can't find it on Bioconda. Maybe @nservant / @mahesh-panchal / @remiolsen or anyone else in the community who has experience with these tools could suggest a better alternative than just bundling the entire source in the pipeline code?

<details>
<summary>Original comment</summary>

Additionally, the Juicer process does not have a `process.container` defined which is required when running on most cloud environments. Even if we end up sticking with the `.jar` approach above, a basic base container still needs to be defined - even if it's just got the basic OS required. For example, something like `quay.io/biocontainers/java-jdk:8.0.112--1`. See for example the [custom python script in the template](https://github.com/nf-core/tools/blob/b48bce26b7da46cb71e4a8f78b68d9ceac579ca6/nf_core/pipeline-template/modules/local/get_software_versions.nf#L11-L16).

</details>

_Edit - I just saw that you already handled part 2 of this; the base container since I first looked, almost exactly as I suggest here in the little hidden comment above 😅 Perfect!_

https://github.com/nf-core/hicar/blob/08910a81d26f8f5ff73d1565ae1eb2021ea7302a/modules/local/cooler/juicer.nf#L15-L20

## R packaging

<details>
<summary>Original comment</summary>

The [`install_packages.r` script](https://github.com/nf-core/hicar/blob/master/bin/install_packages.r) is used in several processes which seems to install required R packages into the local environment. Whilst I guess that this works in a lot of scenarios, it will not work offline. We require all nf-core pipelines to work in offline environments.

Looking at where this script is called, most of the things it's installing seem to be bioconductor packages. This is good news, as all bioconductor packages are hosted in bioconda, so we have ready-made biocontainers for them all. Others are standard enough to be available in conda-forge and I guess we can whip up a biocontainer fairly easily. Quick search:

* [ATACseqQC](https://bioconda.github.io/recipes/bioconductor-atacseqqc/README.html#package-bioconductor-atacseqqc) (bioconda)
* [data.table](https://anaconda.org/conda-forge/r-data.table) (conda-forge)
* [VGAM](https://anaconda.org/conda-forge/r-vgam) (conda-forge)
* [MASS](https://anaconda.org/conda-forge/r-mass) (conda-forge)
* [chippeakanno](https://bioconda.github.io/recipes/bioconductor-chippeakanno/README.html#package-bioconductor-chippeakanno) (bioconda)
* ..and so on..

So I think that this script should be deleted and instead each process that uses it should use an existing distribution with the packages that it requires. If you need help with this then please reach out on Slack.

</details>

_Edit: I just read your comment https://github.com/nf-core/hicar/pull/1#discussion_r741521282 and realise that I somehow missed that this installation script is _in addition to_ the bundled packages. Which is brilliant news!_ 🎉

https://github.com/nf-core/hicar/blob/08910a81d26f8f5ff73d1565ae1eb2021ea7302a/modules/local/atacreads/atacqc.nf#L13-L18

Ok so then my only remaining question on this is whether we actually need this script at all. You say:

> Just in case if user is not using container and it happened they forget to install the package locally. It will not just check the availability of the package.

But I don't see when this would ever be needed apart from if the user is not using _any_ software packaing (conda, docker etc). Running like this with entirely self-installed software is highly discouraged as it breaks the reproducibility of the pipeline.

As such I would suggest that we remove this mechanism, as I think it adds confusion (for me anyway!) and complexity, and hopefully shouldn't really ever be needed. What do you think?

ewels commented on CHANGELOG.md:
```diff
@@ -3,14 +3,24 @@
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
-## v1.0dev - [date]
+## v1.0.2dev - [11/02/2021]
```
We deleted the release, so the changelog must start again at release v1.0.0 - it should reflect the available releases and version numbers in the GitHub repo.

jianhong replied:
Should I just delete the version change history and merge all the change as v1.0.0dev, right?

ewels replied:
Yup! Or as we're hoping to go for a release right away, maybe 1.0 not dev.

nservant commented:
Hi,
Indeed, Juicer is not in conda so far ... that's why I did not use it in the `nfcore/hic` pipeline.
And I think that the authors do not plan to make it available through conda ...

ewels commented on nextflow.config:
```diff
@@ -161,7 +210,7 @@ manifest {
     mainScript      = 'main.nf'
     nextflowVersion = '!>=21.04.0'
-    version         = '1.0dev'
+    version         = '1.0.2dev'
```
We deleted the initial release, so please reset back to `1.0dev`

ewels commented on modules/local/atacreads/macs2.nf:
```diff
@@ -0,0 +1,50 @@
+// Import generic module functions
```
Continuing discussion https://github.com/nf-core/hicar/pull/1#discussion_r741521686

> Because the central one does not support BED format. This can be changed once the central one have BED format support.

Can we update the version in nf-core/modules and use that here instead of having a local module? Maybe there is already something in progress for this, but I couldn't find it when I just did a quick search..

jianhong replied:
That will be great if we have that one. Once it is possible, I move to nf-core/modules.

ewels replied:
@jianhong - are you able to make a pull request to nf-core/modules with this local module then, so that the central copy is updated with your additions?

jianhong replied:
I created a pr just now.

ewels replied:
x-ref https://github.com/nf-core/modules/pull/1038

ewels commented on docs/output.md:
```diff
@@ -37,32 +100,104 @@ The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes d
+</details>
+
+The TADs and loops are called by [MAPS](https://pubmed.ncbi.nlm.nih.gov/30986246/).
+MAPS is a model-based analysis of long-range chromatin interactions.
```
https://github.com/nf-core/hicar/pull/1#discussion_r741483454

If it's possible to add a little more content to this documentation then that would be great. Pretend that you're describing the outputs to someone totally new to this data type and analysis type.

For an example of another pipeline with really nice docs in this file, check out the atacseq pipeline: https://nf-co.re/atacseq/dev/output#call-peaks

jianhong replied:
I will try to add more detailed descriptions.

ewels commented on assets/multiqc_config.yaml:
```diff
@@ -9,3 +9,187 @@ report_section_order:
+
+            *TSS enrichment score* is a ratio between aggregate distribution of reads centered on TSSs
+            and that flanking the corresponding TSSs.
+            TSS score = the depth of TSS (each 100bp window within 1000 bp each side) / the depth of end flanks (100bp each end).
```
Please consider breaking some of these long descriptions out into a `description` and also `helptext`. Then you can use bullet points and more formatting to make it more readable.

See https://multiqc.info/docs/#step-5-create-report-sections - let me know if this doesn't work with Custom Content.

jianhong replied:
Thank you for the help link. I will update long description into a `helptext`.

jianhong replied:
I have trouble in showing the `helptext` in QC report html. I will push the change I made. Please correct me anything I missed.

ewels commented on README.md:
````diff
@@ -49,10 +56,17 @@ On release, automated continuous integration tests run the pipeline on a full-si
 
     ```console
-    nextflow run nf-core/hicar -profile <docker/singularity/podman/shifter/charliecloud/conda/institute> --input samplesheet.csv --genome GRCh37
+    nextflow run nf-core/hicar -profile <docker/singularity/podman/shifter/charliecloud/conda/institute> -c profile.config
````
https://github.com/nf-core/hicar/pull/1#discussion_r741471586

I don't think that there are loads of pipeline parameters for this pipeline, right? So maybe it's ok to keep the docs here mentioning just the essential bare-minimum command line flags to keep things as simple as possible.

jianhong replied:
I changed it to `-c profile.config`

ewels replied:
More what I meant was to drop the config file completely and use command line flags instead, to make it as simple and obvious as possible. Taking the pipeline-specific parameters that you have in the example config file then:

```suggestion
    nextflow run nf-core/hicar -profile <docker/singularity/podman/shifter/charliecloud/conda/institute> \
        --input samples.csv \   # Input data
        --qval_thresh 0.01 \    # Cut-off q-value for MACS2
        --genome GRCh38 \       # Genome Reference
        --mappability /path/mappability/bigWig/file  # Provide mappability to avoid memory intensive calculation
```

ewels commented on docs/usage.md:
````diff
@@ -21,43 +25,88 @@ You will need to create a samplesheet with information about the samples you wou
 
 An [example samplesheet](../assets/samplesheet.csv) has been provided with the pipeline.
 
+### Config file
+
+Users can also run the pipeline by provide a config file by `-c` parameter (see `-c` section below).
+
+```bash
+nextflow run nf-core/hicar -c 'path/to/the/profile/file' -profile docker
+```
+
+In the config file, the design table, and more options can be defined.
+
+Here is an example profile.config file for running nf-core/hicar on slurm cluster.
+
+```groovy
+// submit by slurm
+process.executor = "slurm"
+process.clusterOptions = "-J nextFlowHiCAR"
+
+params {
+    config_profile_name = 'profile name here'
+    config_profile_description = 'profile description here'
+    // report email
+    email = 'users@email.addr'
+    // resources
+    max_cpus   = 8
+    max_memory = 60.GB
+    max_time   = 48.h
````
https://github.com/nf-core/hicar/pull/1#discussion_r741471586

Please avoid adding pipeline-specific documentation about general topics such as running on specific hardware (a cluster, slurm etc). It's better to point people to the central documentation on these things, so that we have a single location to keep up to date.

jianhong replied:
I agree. This section will be removed.

ewels commented on README.md:
```diff
@@ -16,20 +16,27 @@
 
-<!-- TODO nf-core: Write a 1-2 sentence summary of what data the pipeline is for and what it does -->
-**nf-core/hicar** is a bioinformatics best-practice analysis pipeline for This pipeline analyses data for HiCAR data, a robust and sensitive multi-omic co-assay for simultaneous measurement of transcriptome, chromatin accessibility and cis-regulatory chromatin contacts..
+**nf-core/hicar** is a bioinformatics best-practice analysis pipeline for HiC on Accessible Regulatory DNA (HiCAR) data, a robust and sensitive assay for simultaneous measurement of chromatin accessibility and cis-regulatory chromatin contacts. Unlike the immunoprecipitation-based methods such as HiChIP, PlAC-seq and ChIA-PET, HiCAR does not need antibody available. HiCAR utilize the Assay of Tra ...
```
https://github.com/nf-core/hicar/pull/1#discussion_r741469407

Thanks for the improvement here 👍🏻  Are there any citations for this method? I guess https://www.biorxiv.org/content/10.1101/2020.11.02.366062v1 ?

jianhong replied:
Yes. It is. I will update that.

ewels reviewed (commented):
Many thanks for your updates after my review on #1 @jianhong 👍🏻 I've gone back through my comments and copied over any that I thought were still ongoing and not yet completely resolved.

ewels commented:
Thanks @nservant 👍🏻  I had a feeling that we had discussed this before. Any ideas on how we can resolve this? Are the authors explicitly against having it in bioconda? As anyone can add any package there, so we don't need to rely on them doing it - we can add it ourselves. Alternatively I wonder if we can come up with some other solution..

nservant commented:
There is a Docker if it can help ; https://github.com/theaidenlab/Juicer-Docker
I do not think they are really against conda ... I'm just wondering if there is any technical issue actually because If I'm correct part of the Juicer code requires cuda GPU, so I guess it easier to build a container like Docker ?

mahesh-panchal commented:
> There is a Docker if it can help ; https://github.com/theaidenlab/Juicer-Docker I do not think they are really against conda ... I'm just wondering if there is any technical issue actually because If I'm correct part of the Juicer code requires cuda GPU, so I guess it easier to build a container like Docker ?

Their dockerfile doesn't do anything complex: https://github.com/theaidenlab/Juicer-Docker/blob/master/Dockerfile
I don't think they even have GPU tools installed in the docker container (but I could be wrong).

ewels commented:
I wonder if it's worth having a go then.. Maybe worth making an issue on [their repo](https://github.com/aidenlab/juicer about it. I can only find one semi-relevant issue on either the bioconda or juicer repos: https://github.com/aidenlab/juicer/issues/91

mahesh-panchal commented:
That issue there shows what I thought. The GPU aspect isn't supported in the container. Doing what the Dockerfile does appears to be potentially trivial, but if someone wants to use the GPU functionality, then it's likely going to open up a whole can of worms for whoever makes the conda package.

mahesh-panchal commented:
I had a look over the [bioconda contributions guidelines](https://bioconda.github.io/contributor/guidelines.html). 
I guess one could say that the HiCCUPs package is not included or needs to be installed separately (e.g. GATK is not licensed for distribution so needs to be registered; i.e. there's a mechanism for this). 
This is not something I think I have time to write though.

ewels commented:
Ok, so do you think that bundling the jar file as we are here is the best solution for now then @mahesh-panchal?

mahesh-panchal commented:
Hmm. Another option is to stage it as one might do a database file. 
It would work better with the local code anyway in my opinion:
```
java ${params.juicer_jvm} -jar $projectDir/bin/juicer_tools_1.22.01.jar pre \
        -r \$res \
        $options.args --threads $task.cpus ${gi}.sorted ${prefix}.${meta.bin}.hic $chromsize
```
Replace `$projectDir/bin/juicer_tools_1.22.01.jar` with a variable from the input declaration, e.g. `juicer_tools_jar`, and then use a `params.juicer_tools_jar = "https://github.com/aidenlab/Juicebox/releases/download/v2.13.06/juicer_tools_2.13.06.jar"`

mahesh-panchal commented:
I think it's best to leave `bin` only with custom scripts. If something is downloadable, then stage it like any other file if it's not packaged in the usual ways.

mahesh-panchal commented on modules/local/cooler/juicer.nf:
```diff
@@ -0,0 +1,52 @@
+        fi
+    done
+    res=\$(join_by , \${res[@]})
+    java ${params.juicer_jvm} -jar $projectDir/bin/juicer_tools_1.22.01.jar pre \
```
The `${params.juicer_jvm}` should come in as an input value. 
e.g.
```
   input:
    tuple val(meta), path(gi)
    path chromsize
    val juicer_jvm_params
```
and then just add to the workflow, the extra params:
```
JUICER( DUMPINTRAREADS.out.gi, chromsizes, params.juicer_jvm_params )
```

jianhong replied:
Thank you @mahesh-panchal . I will drop juicer_tools and pass it as a input value also.

ewels replied:
Is this resolved now?

nservant commented:
If you use `params.juicer_tools_jar = "https://github.com/aidenlab/Juicebox/releases/download/v2.13.06/juicer_tools_2.13.06.jar"`, will the jar be downloaded on-the-fly ?

mahesh-panchal commented:
> If you use `params.juicer_tools_jar = "https://github.com/aidenlab/Juicebox/releases/download/v2.13.06/juicer_tools_2.13.06.jar"`, will the jar be downloaded on-the-fly ?

It will be staged locally as soon it is needed for a process, and then symlinked to the workdir, yes. 

This also makes it more important to use the tools version reporting if it has it. I see the Juicer process uses an echo, so perhaps strip the version number from the filename instead.

mahesh-panchal commented:
Has the multiple software versions thing been rolled out now? I see the txt method of versions reporting instead of the yaml.

jianhong commented:
I update the `dev` branch according the comments on this PR. Let me know if I missed anything.

ewels commented:
PR #5 merged with your updates @jianhong - then I made and immediately merged #6 from `dev` into `master`. They should now show here as this PR is based on that branch.

ewels commented:
Nice, nearly there I think! Unresolved comments that I can see above:

- [ ] https://github.com/nf-core/hicar/pull/2#issuecomment-958994943
- [ ] https://github.com/nf-core/hicar/pull/2#discussion_r742962354
- [ ] https://github.com/nf-core/hicar/pull/2#discussion_r741702623 (and generally checking on versions everywhere - I think on reflection that we should bump to version `1.0`)

Then just need:

- [ ] At least one more review from someone else, preferably two
    * @mahesh-panchal started, so just needs a ✅ from him
    * Preferably another review from someone. @nservant if you have time? Or a DSL2 maestro like @JoseEspinosa or @FriederikeHanssen or @drpatelh or someone...
- [ ] Test running the pipeline offline (anyone)

ewels commented:
ok first question resolved on Slack: https://nfcore.slack.com/archives/CJRH30T6V/p1636040667478100

@mahesh-panchal (PR comment above)
> Has the multiple software versions thing been rolled out now? I see the txt method of versions reporting instead of the yaml.

@drpatelh (slack): 
> All of the modules on nf-core/modules have been updated with the new, multiple versions syntax. So I guess it's fine to use the old syntax for this release and then maybe update in one go for the next release with the new DSL2 syntax too.
> Just means you won't be able to ask for any updates to modules on that PR if they are using the old syntax.

ewels commented:
It would also be good to write some documentation about Juicer, and how to run with different versions (eg. CUDA etc).

FriederikeHanssen commented:
Ran the pipeline on our cluster with 
`nextflow run nf-core/hicar -r dev -profile test,cfc`

2 processes errored: 
```
NFCORE_HICAR:HICAR:IGV 
NFCORE_HICAR:HICAR:BIOC_CHIPPEAKANNO_MAPS
```

Error is ignored and the pipeline finishes. The error messages are:

```
NFCORE_HICAR:HICAR:IGV 

WARNING: Skipping mount /var/singularity/mnt/session/etc/resolv.conf [files]: /etc/resolv.conf doesn't exist in container
work/14/aeb11b203966d831264f342167618b/.command.sh: line 5: paste: command not found
```

mahesh-panchal commented:
What's the usual practice here with the MAP's pipeline (https://github.com/ijuric/MAPS), where some scripts have been included here in bin (e.g. `bin/MAPS.py`, ` bin/MAPS_peak_formatting.r`). 

Also why does `bin/annopeaks.r ` need to be in bin if it's in a conda package? 

It looks like a few scripts have been pulled in from various places, but I'm not sure where. I understand @jianhong authored some of those scripts, but it seems these should all be packaged or referenced by source. For the time being, perhaps it would be better to do like juicer, and download the scripts on the fly, so one knows where they came from. Otherwise one needs to maintain them here too.

jianhong commented:
> What's the usual practice here with the MAP's pipeline (https://github.com/ijuric/MAPS), where some scripts have been included here in bin (e.g. `bin/MAPS.py`, ` bin/MAPS_peak_formatting.r`).

The reason for this is because there are some steps are not available in their scripts for example automatic detection of chromosome names. I am not sure what is the best practice for this. Should I ask a pull request in their github or not.

> 
> Also why does `bin/annopeaks.r ` need to be in bin if it's in a conda package?

`annopeaks.r` is not available in conda. It call chippeakanno package to do annotation. It is a help function to collect the annotation resources. Similar things is for `trackviewer.r`

jianhong commented:
> It would also be good to write some documentation about Juicer, and how to run with different versions (eg. CUDA etc).

I will update this.

jianhong commented:
> Ran the pipeline on our cluster with `nextflow run nf-core/hicar -r dev -profile test,cfc`
> 
> 2 processes errored:
> 
> ```
> NFCORE_HICAR:HICAR:IGV 
> NFCORE_HICAR:HICAR:BIOC_CHIPPEAKANNO_MAPS
> ```
> 
> Error is ignored and the pipeline finishes. The error messages are:
> 
> ```
> NFCORE_HICAR:HICAR:IGV 
> 
> WARNING: Skipping mount /var/singularity/mnt/session/etc/resolv.conf [files]: /etc/resolv.conf doesn't exist in container
> work/14/aeb11b203966d831264f342167618b/.command.sh: line 5: paste: command not found
> ```

Thank you for testing. This will be fixed soon.

jianhong commented:
> It looks like a few scripts have been pulled in from various places, but I'm not sure where. I understand @jianhong authored some of those scripts, but it seems these should all be packaged or referenced by source. For the time being, perhaps it would be better to do like juicer, and download the scripts on the fly, so one knows where they came from. Otherwise one needs to maintain them here too.

Will it be good to add some documentation to say where the code come from? I modified the scripts to make it strong to handle some errors. I will create PRs for those change.

mahesh-panchal commented:
No, I misunderstood. I didn't look step for step at the scripts. I didn't realise they had been modified. Since they are custom scripts then they belong there in bin. 
Perhaps comment the scripts though to identify the source, the date you downloaded it, and what modifications you made so if in future the tools are updated in their respective repositories, it makes it easier to update the ones in bin ( if you've already done this then sorry. I haven't looked over everything yet).

jianhong commented:
> No, I misunderstood. I didn't look step for step at the scripts. I didn't realise they had been modified. Since they are custom scripts then they belong there in bin. Perhaps comment the scripts though to identify the source, the date you downloaded it, and what modifications you made so if in future the tools are updated in their respective repositories, it makes it easier to update the ones in bin ( if you've already done this then sorry. I haven't looked over everything yet).

My apologize I did not document it well. I will fix this and make it easier for future maintain.

mahesh-panchal commented on modules/local/atacreads/dumpreads.nf:
```diff
@@ -0,0 +1,38 @@
+    path  "*.version.txt"                                                , emit: version
+
+    script:
+    def software = "awk"
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+    def outdir   = "${meta.id}"
+    """
+    mkdir -p $outdir
+    cat $bed | zcat | awk -v setname=${prefix} -v outdir=${outdir}  -F \$"\t"  'BEGIN {OFS=FS};{print \$1,\$2,\$3 > outdir"/"setname"."\$1".shrt.vip.bed"}'
+
+    echo \$(awk --version 2>&1) | sed -e "s/GNU Awk //g; s/, API.*\$//" > ${software}.version.txt
```
```suggestion
    def software = "awk"
    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
    """
    zcat $bed | \\
        awk -F "\t" 'BEGIN { OFS=FS } { print \$1,\$2,\$3 > ${prefix}"."\$1".shrt.vip.bed" }'

    echo \$(awk --version 2>&1) | sed -e "s/GNU Awk //g; s/, API.*\$//" > ${software}.version.txt
```
To make this a bit more readable for me, I trimmed down the command a bit to reduce the number of variables.
The output path above would also need to be updated too. 
What was the $ in `-F \$"\t"` doing?

jianhong replied:
`-F \$"\t"` should be a typo. I will correct this.

mahesh-panchal commented on subworkflows/local/cooler.nf:
```diff
@@ -0,0 +1,56 @@
+    // dump long.intra.bedpe for each group for MAPS to call peaks
+    COOLER_DUMP(COOLER_MERGE.out.cool).bedpe | DUMPINTRAREADS
+    if(juicer_tools_jar){
+        JUICER(DUMPINTRAREADS.out.gi.combine(juicer_tools_jar), chromsizes, juicer_jvm_params)
```
Why do you `combine` the jar file here with the input?

jianhong replied:
Indeed I am not sure why. If I do not combine them, only the first one will be run. With `combine` all `gi` will be run.

mahesh-panchal replied:
Ah, nextflow has two types of channels. Queue type and value type channels. If input comes in as a queue type channel, then the each element of the queue is consumed and another process cannot reuse it. You need to specify the juicer_tools_jar as a value type channel so it is reused. 
Specifically this line in `hicar.nf`:
```
ch_juicer_tools        = params.juicer_tools_jar ? Channel.fromPath(params.juicer_tools_jar, checkIfExists: true) : Channel.empty()
```
Makes a queue type channel `fromPath`.
I think you can do:
```
ch_juicer_tools        = params.juicer_tools_jar ? file(params.juicer_tools_jar, checkIfExists: true) : Channel.empty()
```
to make it a value type channel (test with a toy example, otherwise the `collect` operator also makes a value type channel).

jianhong replied:
Thank you for explain this. It puzzled me for long time. I will change it to value channel.

jianhong commented on modules/local/circos/circos_prepare.nf:
```diff
@@ -0,0 +1,37 @@
+
+    script:
+    """
+    circos.r \\
```
yes, they will be used always together. I separate them to make sure the container will be ready for each step. And they will be connected in `subworkflow`.

mahesh-panchal replied:
I'm not sure I understand. The container is always prepared before running the process anyway. 
Also it's better practice to bundle short tasks into a larger task, particularly if they're related.

jianhong replied:
I agree. I need some time to figure out how to make it possible for the dependencies available in one container.

mahesh-panchal replied:
Oh, I thought it was already part of the same container/tool package. It's been a long time since I used circos. One possibility might be a mulled container, but if they're separate tools then one should keep them separate.

jianhong commented on modules/local/checksums.nf:
```diff
@@ -0,0 +1,58 @@
+    def software = getSoftwareName(task.process)
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+    """
+    touch md5.${prefix}.txt
+    [ ! -f  ${prefix}_1.fastq.gz ] && ln -s ${reads[0]} ${prefix}_1.fastq.gz
+    [ ! -f  ${prefix}_2.fastq.gz ] && ln -s ${reads[1]} ${prefix}_2.fastq.gz
+    gunzip -c ${prefix}_1.fastq.gz > ${prefix}_1.fastq
+    md5.py ${prefix}_1.fastq >>md5.${prefix}.txt
+    gunzip -c ${prefix}_2.fastq.gz > ${prefix}_2.fastq
+    md5.py ${prefix}_2.fastq >>md5.${prefix}.txt
+    if [ "${meta.md5_1}" != "null" ] && [ "${meta.md5_1}" != "" ]; then
+        md5=(\$(md5.py ${prefix}_1.fastq.gz))
+        if [ "\$md5" != "${meta.md5_1}" ]
+        then
+            echo "${meta.id} has checksum ${meta.md5_1}, but we got checksum \$md5!"
+            exit 128
+        fi
+    fi
+    if [ "${meta.md5_2}" != "null" ] && [ "${meta.md5_2}" != "" ]; then
+        md5=(\$(md5.py ${prefix}_2.fastq.gz))
+        if [ "\$md5" != "${meta.md5_2}" ]
+        then
+            echo "${meta.id} has checksum ${meta.md5_2}, but we got checksum \$md5!"
+            exit 128
+        fi
+    fi
+
+    echo \$(python --version) | sed 's/Python //'> python.version.txt
```
I was struggling with container for md5sum. I did not find a good one for this. Do you have any suggestion? Do you think gnu coreutils will work? I also don't want to use custom python script here.

mahesh-panchal replied:
This container contains the `md5sum` utility. It may even be in the biocontainers base image.

jianhong commented on modules/local/atacreads/shiftreads.nf:
```diff
@@ -0,0 +1,38 @@
+    def software = "awk"
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+    """
+    zcat < $pair | \\
+    awk 'BEGIN {OFS="\t"} ;  /^[^#]/ { if (\$7 == "+") {\$5 = \$5 + 4} else if (\$7 == "-") {\$5 = \$5 - 5};  print \$4, \$5, \$5+1, "*", "0", \$7}' | \\
+    sort -k1,1 -k2,2n | uniq  | gzip -nc > ${prefix}.R2.ATAC.bed.gz
```
Thank you for all this. I will update them.

jianhong commented on modules/local/bioc/trackviewer.nf:
```diff
@@ -0,0 +1,40 @@
+    script:
+    prefix   = options.suffix ? "${options.suffix}${bin_size}" : "diffhic_bin${bin_size}"
+    """
+    trackviewer.r -g "${gtf}" -s "${chrom_sizes}" -x "${restrict}" -r $bin_size -o "${prefix}" $options.args
+
+    # *.version.txt files will be created in the rscripts
```
I have trouble in understand this part. Do you mean we will drop the scripts saved in `bin` folder and put the scripts in the `script:` sections?

mahesh-panchal replied:
Yes, I think that makes it easier to maintain since the scripts are custom anyway.

jianhong commented on modules/local/igv.nf:
```diff
@@ -0,0 +1,43 @@
+    def sampleLabel = name.join(' ')
+    def tracks      = track.join(' ')
+    """
+    n=($sampleLabel)
```
I simply use a `for` loop to replace `paste` here. The updates is in pull [#7](https://github.com/nf-core/hicar/pull/7)

jianhong replied:
@mahesh-panchal I followed your suggestion to remake the channel and updated the igv module. I feel your code are more strong at cross-platform.

jianhong commented:
Should I still keep the `bin/scrape_software_versions.py` and `modules/local/get_software_versions.nf`?

mahesh-panchal commented:
> Should I still keep the `bin/scrape_software_versions.py` and `modules/local/get_software_versions.nf`?

These are needed if you're using the old method of reporting software versions ( i.e. with txt files ).

jianhong commented:
> > Should I still keep the `bin/scrape_software_versions.py` and `modules/local/get_software_versions.nf`?
> 
> These are needed if you're using the old method of reporting software versions ( i.e. with txt files ).

I tried. The pipeline will work. However, if I removed them, it will report error in current lint even I do not reporting software version by txt files. Maybe this is a potential bug in current lint tool.

mahesh-panchal commented on modules/local/checksums.nf:
```diff
@@ -0,0 +1,60 @@
+    script:
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+    """
+    touch md5.${prefix}.txt
+    [ ! -f  ${prefix}_1.fastq.gz ] && ln -s ${reads[0]} ${prefix}_1.fastq.gz
+    [ ! -f  ${prefix}_2.fastq.gz ] && ln -s ${reads[1]} ${prefix}_2.fastq.gz
+    gunzip -c ${prefix}_1.fastq.gz > ${prefix}_1.fastq
+    md5sum ${prefix}_1.fastq >>md5.${prefix}.txt
+    gunzip -c ${prefix}_2.fastq.gz > ${prefix}_2.fastq
+    md5sum ${prefix}_2.fastq >>md5.${prefix}.txt
+    if [ "${meta.md5_1}" != "null" ] && [ "${meta.md5_1}" != "" ]; then
+        md5=(\$(md5sum ${prefix}_1.fastq.gz))
+        if [ "\$md5" != "${meta.md5_1}" ]
+        then
+            echo "${meta.id} has checksum ${meta.md5_1}, but we got checksum \$md5!"
+            exit 128
+        fi
+    fi
+    if [ "${meta.md5_2}" != "null" ] && [ "${meta.md5_2}" != "" ]; then
+        md5=(\$(md5sum ${prefix}_2.fastq.gz))
+        if [ "\$md5" != "${meta.md5_2}" ]
+        then
+            echo "${meta.id} has checksum ${meta.md5_2}, but we got checksum \$md5!"
+            exit 128
+        fi
+    fi
```
```suggestion
    [ ! -f  ${prefix}_1.fastq.gz ] && ln -s ${reads[0]} ${prefix}_1.fastq.gz
    [ ! -f  ${prefix}_2.fastq.gz ] && ln -s ${reads[1]} ${prefix}_2.fastq.gz
    gunzip -c ${prefix}_1.fastq.gz > ${prefix}_1.fastq
    gunzip -c ${prefix}_2.fastq.gz > ${prefix}_2.fastq
    md5sum ${prefix}_1.fastq ${prefix}_2.fastq > ${prefix}.md5
    if [ ! -z "${meta.md5_1 ?: ''}" ] && [ ! -z "${meta.md5_2 ?: ''}" ]; then
        cat <<-END_CHECKSUM | md5sum -c
        ${meta.md5_1}  ${prefix}_1.fastq
        ${meta.md5_2}  ${prefix}_2.fastq
        END_CHECKSUM
    fi
```
It's possible to shorten the code a little by piping a heredoc to the `md5sum -c` option, but I'm not sure if you intended to do each file separately on purpose.

jianhong replied:
For hicar, always Paired End. Your code will work great! For general usage, it will be better to check it one by one.

jianhong commented on bin/MAPS_regression_and_peak_caller.r:
```diff
@@ -0,0 +1,390 @@
+# Author: [Ivan Juric](https://github.com/ijuric)
+# File: MAPS_regression_and_peak_caller.r
+# Source: https://github.com/ijuric/MAPS/blob/master/bin/MAPS/MAPS_regression_and_peak_caller.r
+# Data: 11/08/2021
```
I will add this.

jianhong commented on bin/restriction_cut_multipleenzyme.py:
```diff
@@ -0,0 +1,134 @@
+
+#########################################
+# Author: Yunjiang Qiu <serein927@gmail.com>
+# Modified by: Armen Abnousi
```
This step is time consuming. I modified the source with multiprocessing to improve the efficiency. 
I will put the download date and commit to the header.

jianhong commented on docs/output.md:
````diff
@@ -37,32 +103,160 @@ The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes d
+```
+
+[Juicebox](https://github.com/aidenlab/Juicebox) can be used to view the `.hic` file.
+`.hic` file is the second choices for visualization in case installation of `higlass-manage` is impossible.
````
Thank you for all the revising about the grammar.

jianhong commented on modules/local/atacreads/mergepeak.nf:
```diff
@@ -0,0 +1,39 @@
+    script:
+    def software = "bedtools"
+    """
+    cat *.narrowPeak | cut -f1-3 | sort -k1,1 -k2,2n |
+    bedtools merge $options.args \\
+        -i stdin > merged_peak.bed
```
Thank for catch this up. I will fix it.

jianhong commented on modules/local/atacreads/shiftreads.nf:
```diff
@@ -0,0 +1,44 @@
+    def software = "awk"
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+    """
+    cat $pair | \\
+        zcat | \\
+        awk 'BEGIN {OFS="\t"};  /^[^#]/ { if (\$7 == "+") {\$5 = \$5 + 4} else if (\$7 == "-") {\$5 = \$5 - 5};  print \$4, \$5, \$5+1, "*", "0", \$7}' | \\
+        sort -k1,1 -k2,2n | \\
+        uniq | \\
+        gzip -nc > ${prefix}.R2.ATAC.bed.gz
```
why I always keep using `cat input | zcat` is because it will not work if I use `zcat input` on mac os. In mac OS, it will always add `.Z` to the end of the filename.

mahesh-panchal replied:
Understood, but I think the default assumption is that the workflows are intended to be reproducible. As such they should always use a container technology or conda. 

@ewels What is the base assumption of software packaging?

mahesh-panchal replied:
Actually, I just realized, that would mean the behaviour is inconsistent. If using `cat` only bypasses the error message, `zcat` should still be performing the same operation regardless if it reads from a file or from a pipe.

jianhong replied:
I will try to use different container to fix this issue.

mahesh-panchal replied:
Sorry, I just want be clear here. We should not use `zcat` from MacOS at all here since it's working on `.Z` files. Even though `cat <file> | zcat | ` does not report an error, it means it's still decompressing `.Z` from stdin instead. It is better then, that we use `zcat <file>` so an error is reported if the wrong `zcat` is used.

jianhong replied:
If you use docker/singularity, it is true that the pipeline will use the function in container. It will be different for `conda`. `conda` simply export the PATH. And most of the container for `conda` will not include `zcat`. In MacOS, `zcat` working on `.Z` files. By pipe `cat <file> | zcat` will avoid the filename issue. And the `filename` should not contain wildcard. The performance may be a little different but the output will be same. Correct me if I am not correct.

mahesh-panchal replied:
The problem is not a filename issue. The problem is that MacOS `zcat` is trying to uncompress Zip archive data format, which is not the same as gzip. Piping to `zcat` on MacOS then only hides the problem, but still uses Zip archive data decompression instead of gzip decompression.

mahesh-panchal replied:
Maybe a better solution is to use `gunzip -c` instead of `zcat`.

jianhong replied:
I will use `gunzip -c` instead. Thank you.

mahesh-panchal commented on subworkflows/local/preparegenome.nf:
```diff
@@ -0,0 +1,155 @@
+    /*
+     * Calculate effective genome sizes
+     */
+    gs = ch_fasta.size() * 0.78
```
Channels do not have a `size` operator. What is happening here?
Where does this 0.78 number come from?

jianhong replied:
I will remove this.

mahesh-panchal commented on subworkflows/local/preparegenome.nf:
```diff
@@ -0,0 +1,155 @@
+    } else {
+        //genome size remove all N then * 78%
+        gs = 0.0
+        ch_fasta.withReader{
```
Again here, channels do not have a `withReader` operator. This only works on files.

jianhong replied:
Sorry. I forget to fix this issue.

mahesh-panchal reviewed (commented):
What is the purpose of the ` modules/nf-core/modules/gunzip/test.txt.gz` ?

jianhong commented on modules/local/hipeak/prepare_counts.nf:
```diff
@@ -0,0 +1,185 @@
+            if(file.exists(fn)){
+                pc <- readRDS(fn)
+            }else{
+                NULL
```
This code is used to speed up the IO process. To avoid to load the whole huge file into memory during whole process, I first split the files by chromosomes and then load the files when required.

mahesh-panchal replied:
Sorry, I specifically meant the `NULL`. It's not assigned to anything, or do I miss where it is assigned?

jianhong replied:
Oh, sorry for the short script. That means it will return a NULL. I will fix this by assign it to variable to make it readable.

jianhong commented on modules/local/maps/reformat.nf:
```diff
@@ -0,0 +1,40 @@
+    """
+    MAPS_peak_formatting.r $bin_size ${peak}
+
+    cat <<-END_VERSIONS > versions.yml
+    ${getProcessName(task.process)}:
+        MAPS: 1.1.0
+    END_VERSIONS
+    for i in \$(ls *.version.txt); do
+    echo "    \${i%.version.txt}: \$(<\$i)" >> versions.yml
+    done
```
I will fix this.

jianhong commented:
I created a new PR for above comment.

mahesh-panchal commented:
> What is the purpose of the ` modules/nf-core/modules/gunzip/test.txt.gz` ?

jianhong commented:
> > What is the purpose of the ` modules/nf-core/modules/gunzip/test.txt.gz` ?

I am not sure. I installed it by nf-core tools. I will ask it in slack channel.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
update ChangeLog.md

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
try to avoid the tracedir issue.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

mahesh-panchal reviewed (approved):
So apart from the `modules/nf-core/modules/gunzip/test.txt.gz` file, I think it's ok. I couldn't see any other big issues.

jianhong committed:
use hdf5 to improve the IO efficency.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
try to fix the issue `No such variable: gunzip` after updated gunzip module

jianhong committed:
Merge pull request #11 from jianhong/dev

update modules and improve the IO efficiency.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
Merge pull request #12 from nf-core/dev

merge Dev to Master

jianhong committed:
update for h5 file locking.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
update h5 file locing for trackViewer.nf

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
make hicar.nf more readable.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
add unit test for local `subworkflow`

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
update pytest_software.yml

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
fix the duplicated mapping key for pytest

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
update pytest_software.yml

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
update pytest_software.yml

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
reduce several test modules from test_full.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
replace bedtool sort by coreutils sort.

jianhong committed:
fix the white space of modules.json.

jianhong committed:
fix the bug introduce by mistyping in DIFF_HIPEAK.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
update the test for hipeak.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

nf-core-bot committed:
Template update for nf-core/tools version 2.2

jianhong committed:
simplify the fragment peak calling by call peaks only for fragmentation sites

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
fix the bug in calling peaks for `pospoisson regression`

jianhong committed:
Merge remote-tracking branch 'origin/dev'

mahesh-panchal commented on CHANGELOG.md:
```diff
@@ -3,14 +3,55 @@
+- fix the empty output of PCA analysis for `multiQC`.
+- replace `zcat` by `gunzip -c`
+- update documentation of readme
+- fix the wrong estimated gonme size for unsupported genome.
```
```suggestion
- fix the wrong estimated genome size for unsupported genome.
```

mahesh-panchal commented on conf/modules.config:
```diff
@@ -22,11 +22,368 @@
 
 params {
     modules {
+        'checksums'{
```
Will you update to the latest syntax version? The one that doesn't use params.modules?

jianhong replied:
@mahesh-panchal I am working on this. Hope I will update the pipeline soon.

jianhong replied:
@mahesh-panchal I already update the latest syntax version. Now it is ready for review.

jianhong committed:
Merge remote-tracking branch 'downstream/nf-core-template-merge-2.2' into merging-template-2.2

jianhong committed:
fix the pytest issue.

jianhong committed:
fix unchanged.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
update ci.yml

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
fix the issues for skipped steps.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
fix a bug that output number in scientific mode for R1 peaks

jianhong committed:
add error-retry for MAPS_CUT and reset the number of multiprocess cores to cores-1 for restriction_cut_multipleenzyme.py

jianhong committed:
roll it back for restriction_cut_multipleenzyme.py

jianhong committed:
change the default publishDir.enabled to false.

jianhong committed:
update restriction_cut_multipleenzyme.py

jianhong committed:
try to add igv module back

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
Merge pull request #13 from jianhong/dev

Update with current template

jianhong committed:
Merge pull request #15 from nf-core/dev

Merge from Dev

jianhong committed:
update errorStrategy to include 140

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
fix the bug for R1 peak calling when the peak is out-of-bound ranges

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
fix the bug for R1 peak calling when the peak is out-of-bound ranges

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
fix lint issue.

jianhong committed:
fix a bug in prepare_counts.nf.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
add chunk for hdf5 file.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
fix typo in nextflow_schema.json.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
adjust the resources for hipeak calling.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
fix the bug for v4c when no available gene annotation.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
try to speed up the prepare_counts step for hipeak.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
fix the wrong input file for prepare_counts

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
add long lable to maps_cut

jianhong committed:
Merge pull request #16 from jianhong/dev

merge upstream Dev

jianhong committed:
handle 0 length of peaks in hipeak.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong commented:
Hi, may I know if there is any new comments for the review? Thank you.

jianhong committed:
Merge pull request #17 from nf-core/dev

merge Dev to master

jianhong committed:
prepare reads count for hipeak by chromosome

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
remove unused binCounts from prepare_counts.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
fix the issue in hipeak test.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
fix the bug prepare_counts does not handle all chromosomes.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
update docs/usage.md

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
handle gis is not a list in prepare_counts.nf

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
try catch the error when parrallel not work.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
decrease the peak_pair_block size.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

mahesh-panchal commented on bin/check_samplesheet.py:
```diff
@@ -1,16 +1,12 @@
 
 def parse_args(args=None):
-    Description = "Reformat nf-core/hicar samplesheet file and check its contents."
+    Description = "Reformat nf-core/ttest samplesheet file and check its contents."
```
Is this a correct change?

mahesh-panchal commented on modules/local/get_software_versions.nf:
```diff
@@ -1,18 +1,11 @@
-
-params.options = [:]
-
 process GET_SOFTWARE_VERSIONS {
```
This process is not needed any more is it? It was replaced with another script no?

jianhong replied:
Yes. It will be removed.

mahesh-panchal commented on nextflow.config:
```diff
@@ -9,14 +9,70 @@
+
+    // HiPeak calling
+    remove_dup                 = false
+    cluster_type               = "SOCK"
```
Where is this parameter used? I couldn't find it easily

jianhong replied:
The cluster_type will be used by the BiocParallel package (indeed it will pass to snow package) to handle multiple thread computation.

mahesh-panchal commented on tests/config/nextflow.config:
```diff
@@ -0,0 +1,77 @@
+    outdir = "output/"
+    publish_dir_mode = "copy"
+    enable_conda = false
+    singularity_pull_docker_container = false
```
old parameter.

mahesh-panchal commented on tests/config/nextflow.config:
```diff
@@ -0,0 +1,77 @@
+params {
+    outdir = "output/"
+    publish_dir_mode = "copy"
```
old parameter

mahesh-panchal commented on tests/config/nextflow.config:
```diff
@@ -0,0 +1,77 @@
+    withName: ASSIGN_TYPE {
+        ext.args    = "--count_cutoff 4 --ratio_cutoff 2 --fdr 1"
+        publishDir  = [
+            path: { "${params.outdir}/HiPeak" }
```
```suggestion
            path: "${params.outdir}/HiPeak"
```
The closure is only needed when you want to access something in task context

jianhong replied:
After some testing, I decide to keep the closure. It will work for change the setting from command line but will not work for changing the setting by configuration file. 

For example, 

`nextflow run nf-core/hicar -profile test_full --outdir output // this one work`

`nextflow run nf-core/hicar -profile test_full -c addition.conf //this one will not work`
whith addition.conf:
params {
   outdir     = "output"
}

mahesh-panchal replied:
Hmm. I forget people keep wanting to supply parameters this way. I think I should submit a change request so it's more inline with what people think should happen.

jianhong committed:
fix the typo of check_samplesheet.py

mahesh-panchal reviewed (commented):
Mostly old syntax that needs to be swapped for new.

jianhong committed:
update the old syntax.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong commented on modules/local/bioc/enrich.nf:
```diff
@@ -0,0 +1,32 @@
+    path "versions.yml"                                , emit: versions
+
+    script:
+    prefix   = task.ext.prefix ? "${task.ext.prefix}${bin_size}" : "diffhic_bin${bin_size}"
```
Thank you for clarify. I will update this.

jianhong committed:
fix a typo in modules.config

jianhong committed:
fix the blank lines in CHANGELOG.md

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
fix a typo in juicer.nf

jianhong committed:
add bracket maker back. Otherwise, the parameters will not change when setting the parameters via user defined config file.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
Improve the memory cost for prepare_counts.nf

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
Merge pull request #18 from jianhong/dev

merging upstream Dev

jianhong committed:
Merge pull request #19 from nf-core/dev

merging from Dev

jianhong committed:
Merge pull request #3 from nf-core/dev

Dev

jianhong committed:
update the MAPS summary table.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
Merge pull request #20 from jianhong/dev

Fix the missing report files of multiQC

jianhong committed:
Merge pull request #21 from nf-core/dev

merging from Dev

jianhong committed:
fix the bug for the out-of-bound interaction in hipeak.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
optimize the efficency of pairs2hdf5.nf

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
change the output folder of filtered pairs.
change the parameter cluster_type to snow_type.
use multiple threads to handle diff_hipeak.

jianhong committed:
update preparegenome.nf try to fix the issue of wrong file path.

jianhong committed:
clean up fastq files after md5 check.

jianhong committed:
add circos plot for MAPS loops

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
update linting.yml to setup-node@v2

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
change it back.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
remove the interactions with distance smaller than 1K for hipeak

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
export MAPS signals to cooler

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
update the linting.yml file

jianhong committed:
try to fix the lint issue

jianhong committed:
fix the typo of filename .yamllint.yml

jianhong committed:
update nf-core modules to meet the requirement of lint.

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
fix the issue for function not found in parallel computing of diff_hipeak.

jianhong committed:
Merge pull request #22 from jianhong/dev

Dev

jianhong committed:
Merge pull request #23 from nf-core/dev

Dev

jianhong committed:
fix the bug with updated wigtobigwig

jianhong committed:
Merge pull request #25 from jianhong/dev

fix the bug with updated wigtobigwig

yxian9 committed:
Update README.md

update readme

jianhong committed:
fix the overflow issue in assign_type.nf

jianhong committed:
Merge remote-tracking branch 'origin/dev'

jianhong committed:
fix the overflow issue in assign-type.nf

jianhong committed:
update the doi link for hicar paper.

jianhong committed:
fix a bug for assign_type.nf for deep looping.

jianhong committed:
Merge pull request #26 from nf-core/dev

Dev

mahesh-panchal commented on docs/usage.md:
```diff
@@ -21,34 +25,39 @@ You will need to create a samplesheet with information about the samples you wou
 ### Full samplesheet
 
-The pipeline will auto-detect whether a sample is single- or paired-end using the information provided in the samplesheet. The samplesheet can have as many columns as you desire, however, there is a strict requirement for the first 3 columns to match those defined in the table below.
+HiCAR is require paired-end sequencing. Both fastq_1 and fastq_2 must be provided in the samplesheet. The samplesheet can have as many columns as you desire, however, there is a strict requirement for the first 4 columns to match those defined in the table below.
```
```suggestion
HiCAR requires paired-end sequencing. Both fastq_1 and fastq_2 must be provided in the samplesheet. The samplesheet can have as many columns as you desire, however, there is a strict requirement for the first 4 columns to match those defined in the table below.
```

jianhong committed:
Merge pull request #27 from jianhong/dev

fix overflow issue for assign_type.nf

mahesh-panchal commented on docs/usage.md:
````diff
@@ -138,85 +147,28 @@ Specify the path to a specific config file (this is a core Nextflow command). Se
-For example, if the nf-core/rnaseq pipeline is failing after multiple re-submissions of the `STAR_ALIGN` process due to an exit code of `137` this would indicate that there is an out of memory issue:
+For example, if the nf-core/hicar pipeline is failing after multiple re-submissions of the `MAPS_CUT` process due to an exit code of `137` this would indicate that there is an out of memory issue:
 
 ```console
````
I think it would be a good idea to leave this example as is, or replace it with a full error message from MAPS_CUT

jianhong replied:
updated the documentation to full error message from MAPS_CUT.

mahesh-panchal reviewed (approved):
Approving for now, and the changes can probably be saved for making a new release

jianhong committed:
Merge pull request #28 from nf-core/dev

Dev

jianhong committed:
Prepare for release.

jianhong committed:
update the version

jianhong committed:
Merge pull request #29 from jianhong/master

Prepare for release.

jianhong committed:
Merge pull request #30 from nf-core/dev

Prepare for release

jianhong committed:
Update conf/modules.config

Co-authored-by: Mahesh Binzer-Panchal <mahesh.binzer-panchal@nbis.se>

jianhong committed:
Update docs/usage.md

Co-authored-by: Mahesh Binzer-Panchal <mahesh.binzer-panchal@nbis.se>

jianhong committed:
Update conf/modules.config

Co-authored-by: Mahesh Binzer-Panchal <mahesh.binzer-panchal@nbis.se>

jianhong committed:
Update conf/modules.config

Co-authored-by: Mahesh Binzer-Panchal <mahesh.binzer-panchal@nbis.se>

jianhong committed:
Update conf/modules.config

Co-authored-by: Mahesh Binzer-Panchal <mahesh.binzer-panchal@nbis.se>

jianhong committed:
Update conf/modules.config

Co-authored-by: Mahesh Binzer-Panchal <mahesh.binzer-panchal@nbis.se>

jianhong committed:
Merge pull request #4 from nf-core/master

update for some comments from @Mahesh Binzer-Panchal

jianhong committed:
Merge remote-tracking branch 'origin/master' into dev

jianhong committed:
update the for the reviewers comments

jianhong committed:
change the version back to 1.0.0

jianhong committed:
add data for awsfulltest

jianhong committed:
change the aws data to published full datasets.

jianhong committed:
Merge pull request #31 from jianhong/dev

add full test data

jianhong committed:
Merge pull request #32 from nf-core/dev

Dev

jianhong committed:
change the full datasets.

jianhong committed:
add hichip support.

jianhong committed:
fix the bug that when annotation failed. 
Update MAPS for new version of VGAM.

jianhong committed:
update test file.

jianhong committed:
add test_hipeak config file.

jianhong committed:
update the test_full_aws_samplesheet.csv to full size files.

jianhong commented on conf/modules.config:
```diff
@@ -1,32 +1,719 @@
+        enabled: false
+    ]
+
+    withName: SAMPLESHEET_CHECK {
```
all process names are quoted.

jianhong commented on conf/modules.config:
```diff
@@ -1,32 +1,719 @@
+        ]
+    }
+    withName: ATACQC {// ATAC QC
+        ext.args    = '.merged.ATAC.bed.gz'
```
changed it to other format to avoid the confusion.

jianhong committed:
update usage documentation

jianhong committed:
Merge pull request #34 from jianhong/dev

Dev

jianhong committed:
Merge pull request #35 from nf-core/dev

merge from Dev

nf-core-bot committed:
Template update for nf-core/tools version 2.3

jianhong committed:
update max_time for test_full_aws.config

jianhong committed:
Merge remote-tracking branch 'origin/dev' into nf-core-template-merge-2.3

jianhong committed:
update for duplicated parameter publish_dir_mode.

jianhong committed:
fix the sample check issue.

jianhong committed:
Merge pull request #37 from jianhong/dev

update max_time for test_full_aws.config

jianhong committed:
Merge remote-tracking branch 'origin/dev_simple' into dev

jianhong committed:
Merge branch 'jianhong:dev' into dev

jianhong committed:
Merge pull request #38 from nf-core/dev

Dev

jianhong committed:
update max_time for test_full_aws.config

jianhong committed:
fix the issue of patch chromosome when run juicer.nf

jianhong committed:
Merge pull request #40 from jianhong/dev

update max_time for test_full_aws.config.

jianhong committed:
Merge pull request #41 from nf-core/dev

Dev

jianhong committed:
add filters of chromosome names for hipeak

nf-core-bot committed:
Template update for nf-core/tools version 2.3.1

jianhong committed:
Merge remote-tracking branch 'downstream/nf-core-template-merge-2.3.1' into dev

jianhong committed:
fix the SyntaxError: Unexpected token.

jianhong committed:
rewrite by prettier

jianhong committed:
update modules for the prettier change.

jianhong committed:
Merge pull request #43 from jianhong/dev

Dev

jianhong committed:
Merge pull request #44 from nf-core/dev

Dev

jianhong committed:
fix the logo path in README.md

jianhong committed:
fix the logo path in README.md

jianhong committed:
Merge pull request #45 from jianhong/dev

Dev

jianhong committed:
Merge pull request #46 from nf-core/dev

Dev

ewels committed:
Fix awsfulltest.yml

Don't overwrite tower-action defaults for pipeline and git sha.

ewels committed:
Update awstest.yml

jianhong committed:
Merge pull request #48 from jianhong/dev

fix-tower-action

jianhong committed:
Merge pull request #49 from nf-core/dev

Dev

jianhong committed:
change the test_full_aws to test_full.

jianhong committed:
update test folder to projectDir

jianhong committed:
Merge pull request #50 from jianhong/dev

change the test_full_aws to test_full.

jianhong committed:
Merge pull request #51 from nf-core/dev

Dev

ewels committed:
Remove max_* params from test_full

nf-core-bot committed:
Template update for nf-core/tools version 2.3.2

jianhong committed:
Merge pull request #52 from nf-core/no-max-resources-test_full

Remove max_* params from test_full

jianhong committed:
Merge remote-tracking branch 'downstream/nf-core-template-merge-2.3.2' into dev

jianhong committed:
change the labels for PREPARE_COUNTS.

jianhong committed:
update the max_memory for defaults.

jianhong committed:
update the juicer_jvm_params for full test.

jianhong committed:
Merge pull request #54 from jianhong/dev

update the max_memory for defaults and juicer_jvm_params

jianhong committed:
Merge pull request #55 from nf-core/dev

Dev

jianhong committed:
adjust resource requirement. add parameter for JUICER to avoid the normalization error.

jianhong committed:
fix the error for test hipeak.

jianhong committed:
remove parallel for sort.

jianhong committed:
Merge pull request #56 from jianhong/dev

adjust resource requirement. add parameter for JUICER to avoid the no…

jianhong committed:
Merge pull request #57 from nf-core/dev

Dev

jianhong committed:
remove the -S parameter for sort in shiftreads and mergereads.nf

jianhong committed:
Merge pull request #58 from jianhong/dev

remove the -S parameter for sort in shiftreads and mergereads.nf

jianhong committed:
Merge pull request #59 from nf-core/dev

Dev

jianhong committed:
fix a bug in post_counts for hipeak when there is empty interactions for one chromosome.

jianhong committed:
Merge pull request #60 from jianhong/dev

fix a bug in post_counts for hipeak when there is empty interactions …

jianhong committed:
Merge pull request #61 from nf-core/dev

Dev

jianhong committed:
add totalLinks parameter for prepare_circos

jianhong committed:
fix the issue if bplapply does not work in diff_hipeak.

jfy133 committed:
Merge pull request #62 from nf-core/master

PR from master to dev

jianhong committed:
prepare for release

jianhong committed:
Merge pull request #63 from jianhong/dev

Dev

jfy133 committed:
Update CHANGELOG.md

jfy133 committed:
Update CHANGELOG.md

jfy133 committed:
Update CHANGELOG.md

jianhong committed:
update CHANGELOG to trigger change in dev.

jfy133 committed:
Update CHANGELOG.md

jfy133 committed:
Merge pull request #67 from jianhong/dev

update CHANGELOG to trigger change in dev.

jianhong committed:
update changelogs to clean up the sentences with 'merging comments'

jianhong committed:
update README and CHANGELOG

jfy133 committed:
Merge pull request #68 from jianhong/dev

Update the changelog and readme files

jianhong committed:
Merge pull request #64 from nf-core/dev

Prepare for release

jfy133 committed:
Add Zenodo DOI after first release

jfy133 committed:
Update Zenodo DOI in citations sectiojn

jfy133 commented:
Can be closed now we have first release!
